File: platform/capture_flutter/ios/Classes/CaptureFlutterPlugin.swift
// capture-sdk - bitdrift's client SDK
// Copyright Bitdrift, Inc. All rights reserved.
//
// Use of this source code is governed by a source available license that can be found in the
// LICENSE file or at:
// https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt

import Capture
import Flutter
import UIKit

public class CaptureFlutterPlugin: NSObject, FlutterPlugin {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "io.bitdrift.capture_flutter",
            binaryMessenger: registrar.messenger()
        )
        let instance = CaptureFlutterPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    private var activeSpans: [String: Any] = [:]

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "start":
            handleStart(call, result: result)
        case "log":
            handleLog(call, result: result)
        case "logScreenView":
            handleLogScreenView(call, result: result)
        case "getSessionId":
            result(Logger.sessionID)
        case "getSessionUrl":
            result(Logger.sessionURL)
        case "getDeviceId":
            result(Logger.deviceID)
        case "createTemporaryDeviceCode":
            handleCreateTemporaryDeviceCode(result: result)
        case "startNewSession":
            Logger.startNewSession()
            result(nil)
        case "getSdkStatus":
            let status = Logger.getSdkStatus()
            let stateName: String
            switch status.initializationState {
            case .notStarted: stateName = "NOT_STARTED"
            case .loaded: stateName = "LOADED"
            case .running: stateName = "RUNNING"
            case .disabled: stateName = "DISABLED"
            @unknown default: stateName = "UNKNOWN"
            }
            result([
                "initializationState": stateName,
                "lastHandshakeTime": status.lastHandshakeTime?.timeIntervalSince1970,
                "lastConfigDeliveryTime": status.lastConfigDeliveryTime?.timeIntervalSince1970,
            ] as [String: Any?])

        case "addField":
            guard let args = call.arguments as? [String: Any],
                  let key = args["key"] as? String,
                  let value = args["value"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing key/value", details: nil))
                return
            }
            Logger.addField(withKey: key, value: value)
            result(nil)
        case "removeField":
            guard let args = call.arguments as? [String: Any],
                  let key = args["key"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing key", details: nil))
                return
            }
            Logger.removeField(withKey: key)
            result(nil)
        case "setEntityId":
            guard let args = call.arguments as? [String: Any],
                  let entityId = args["entityId"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing entityId", details: nil))
                return
            }
            Logger.setEntityID(entityId)
            result(nil)
        case "clearEntityId":
            Logger.clearEntityID()
            result(nil)
        case "logNetworkRequest":
            handleLogNetworkRequest(call, result: result)
        case "logNetworkResponse":
            handleLogNetworkResponse(call, result: result)
        case "startSpan":
            handleStartSpan(call, result: result)
        case "endSpan":
            handleEndSpan(call, result: result)
        case "logReplayScreen":
            handleLogReplayScreen(call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func handleStart(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let apiKey = args["apiKey"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "Missing apiKey", details: nil))
            return
        }
        let apiUrl = args["apiUrl"] as? String ?? "https://api.bitdrift.io"
        let strategyName = args["sessionStrategy"] as? String ?? "fixed"
        let sessionStrategy: SessionStrategy = strategyName == "activityBased"
            ? .activityBased()
            : .fixed()

        let configuration = Configuration(
            // Flutter provides its own wireframe data via logSessionReplayScreen.
            sessionReplayConfiguration: nil,
            apiURL: URL(string: apiUrl)!
        )

        Logger.start(
            withAPIKey: apiKey,
            sessionStrategy: sessionStrategy,
            configuration: configuration
        )
        result(true)
    }

    private func handleLog(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let message = args["message"] as? String,
              let levelStr = args["level"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "Missing params", details: nil))
            return
        }
        let fields = (args["fields"] as? [String: String]) ?? [:]
        switch levelStr {
        case "trace": Logger.logTrace(message, fields: fields)
        case "debug": Logger.logDebug(message, fields: fields)
        case "info": Logger.logInfo(message, fields: fields)
        case "warning": Logger.logWarning(message, fields: fields)
        case "error": Logger.logError(message, fields: fields)
        default: Logger.logInfo(message, fields: fields)
        }
        result(nil)
    }

    private func handleCreateTemporaryDeviceCode(result: @escaping FlutterResult) {
        Logger.createTemporaryDeviceCode { deviceCodeResult in
            switch deviceCodeResult {
            case .success(let deviceCode):
                result(deviceCode)
            case .failure(let error):
                result(FlutterError(code: "DEVICE_CODE_ERROR", message: error.localizedDescription, details: nil))
            }
        }
    }

    private func handleLogScreenView(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let screenName = args["screenName"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "Missing screenName", details: nil))
            return
        }
        Logger.logScreenView(screenName: screenName)
        result(nil)
    }

    private static func parseHTTPURLPath(_ any: Any?) -> HTTPURLPath? {
        guard let map = any as? [String: Any], let value = map["value"] as? String else {
            return nil
        }
        return HTTPURLPath(value: value, template: map["template"] as? String)
    }

    private static func parseHTTPRequestInfo(_ map: [String: Any]) -> HTTPRequestInfo? {
        guard let method = map["method"] as? String, let spanID = map["spanId"] as? String else {
            return nil
        }
        return HTTPRequestInfo(
            method: method,
            host: map["host"] as? String,
            path: Self.parseHTTPURLPath(map["path"]),
            query: map["query"] as? String,
            headers: map["headers"] as? [String: String],
            bytesExpectedToSendCount: (map["bytesExpectedToSendCount"] as? NSNumber)?.int64Value,
            spanID: spanID,
            extraFields: map["extraFields"] as? [String: String]
        )
    }

    private func handleLogNetworkRequest(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let request = Self.parseHTTPRequestInfo(args) else {
            result(FlutterError(code: "INVALID_ARGS", message: "Missing/invalid request", details: nil))
            return
        }
        Logger.log(request)
        result(nil)
    }

    private func handleLogNetworkResponse(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let requestMap = args["request"] as? [String: Any],
              let request = Self.parseHTTPRequestInfo(requestMap),
              let responseMap = args["response"] as? [String: Any],
              let resultStr = responseMap["result"] as? String,
              let durationMs = (args["durationMs"] as? NSNumber)?.doubleValue else {
            result(FlutterError(code: "INVALID_ARGS", message: "Missing/invalid response", details: nil))
            return
        }

        let httpResult: HTTPResponse.HTTPResult
        switch resultStr {
        case "success": httpResult = .success
        case "canceled": httpResult = .canceled
        default: httpResult = .failure
        }

        let error: Error? = (responseMap["error"] as? String).map {
            NSError(domain: "io.bitdrift.capture_flutter", code: 0, userInfo: [NSLocalizedDescriptionKey: $0])
        }

        let response = HTTPResponse(
            result: httpResult,
            host: responseMap["host"] as? String,
            path: Self.parseHTTPURLPath(responseMap["path"]),
            query: responseMap["query"] as? String,
            headers: responseMap["headers"] as? [String: String],
            statusCode: (responseMap["statusCode"] as? NSNumber)?.intValue,
            error: error
        )

        // Native metrics durations are seconds; the Dart API uses milliseconds throughout.
        let metricsMap = args["metrics"] as? [String: Any]
        let metrics: HTTPRequestMetrics? = metricsMap.map { m in
            func seconds(_ key: String) -> TimeInterval? {
                (m[key] as? NSNumber).map { $0.doubleValue / 1000.0 }
            }
            return HTTPRequestMetrics(
                requestBodyBytesSentCount: (m["requestBodyBytesSentCount"] as? NSNumber)?.int64Value,
                responseBodyBytesReceivedCount: (m["responseBodyBytesReceivedCount"] as? NSNumber)?.int64Value,
                requestHeadersBytesCount: (m["requestHeadersBytesCount"] as? NSNumber)?.int64Value,
                responseHeadersBytesCount: (m["responseHeadersBytesCount"] as? NSNumber)?.int64Value,
                dnsResolutionDuration: seconds("dnsResolutionDurationMs"),
                tlsDuration: seconds("tlsDurationMs"),
                tcpDuration: seconds("tcpDurationMs"),
                fetchInitializationDuration: seconds("fetchInitializationMs"),
                responseLatency: seconds("responseLatencyMs"),
                protocolName: m["protocolName"] as? String
            )
        }

        Logger.log(HTTPResponseInfo(
            requestInfo: request,
            response: response,
            duration: durationMs / 1000.0,
            metrics: metrics,
            extraFields: args["extraFields"] as? [String: String]
        ))
        result(nil)
    }

    private func handleStartSpan(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let name = args["name"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "Missing name", details: nil))
            return
        }
        let levelStr = args["level"] as? String ?? "info"
        let level: Capture.LogLevel
        switch levelStr {
        case "trace": level = .trace
        case "debug": level = .debug
        case "warning": level = .warning
        case "error": level = .error
        default: level = .info
        }
        let fields = (args["fields"] as? [String: String]) ?? [:]
        if let span = Logger.startSpan(name: name, level: level, fields: fields) {
            let spanId = "\(ObjectIdentifier(span as AnyObject).hashValue)"
            activeSpans[spanId] = span
            result(spanId)
        } else {
            result(nil)
        }
    }

    private func handleEndSpan(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let spanId = args["spanId"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "Missing spanId", details: nil))
            return
        }
        if let span = activeSpans.removeValue(forKey: spanId) as? Span {
            let success = args["success"] as? Bool ?? true
            if success {
                span.end(.success)
            } else {
                span.end(.failure)
            }
        }
        result(nil)
    }

    private func handleLogReplayScreen(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let screenData = args["screen"] as? FlutterStandardTypedData else {
            result(FlutterError(code: "INVALID_ARGS", message: "Missing screen data", details: nil))
            return
        }
        // logSessionReplayScreen is internal on the iOS SDK (requires @testable import).
        // No public API exists for this yet — placeholder logs replay size as info.
        // Track: expose logSessionReplayScreen on Logger public API.
        Logger.logInfo("_session_replay", fields: ["_replay_size": "\(screenData.data.count)"])
        result(nil)
    }
}

File: platform/capture_flutter/android/src/main/kotlin/io/bitdrift/capture_flutter/CaptureFlutterPlugin.kt
// capture-sdk - bitdrift's client SDK
// Copyright Bitdrift, Inc. All rights reserved.
//
// Use of this source code is governed by a source available license that can be found in the
// LICENSE file or at:
// https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt

@file:Suppress("INVISIBLE_MEMBER", "INVISIBLE_REFERENCE")

package io.bitdrift.capture_flutter

import android.content.Context
import io.bitdrift.capture.Capture
import io.bitdrift.capture.Capture.Logger
import io.bitdrift.capture.CaptureResult
import io.bitdrift.capture.Configuration
import io.bitdrift.capture.IInternalLogger
import io.bitdrift.capture.LogLevel
import io.bitdrift.capture.events.span.Span
import io.bitdrift.capture.events.span.SpanResult
import io.bitdrift.capture.network.HttpRequestInfo
import io.bitdrift.capture.network.HttpRequestMetrics
import io.bitdrift.capture.network.HttpResponse
import io.bitdrift.capture.network.HttpResponseInfo
import io.bitdrift.capture.network.HttpUrlPath
import io.bitdrift.capture.providers.Field
import io.bitdrift.capture.providers.session.SessionStrategy
import io.bitdrift.capture.providers.toFieldValue
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import okhttp3.HttpUrl.Companion.toHttpUrl
import java.util.UUID
import kotlin.time.Duration

class CaptureFlutterPlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var context: Context

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, "io.bitdrift.capture_flutter")
        channel.setMethodCallHandler(this)
        context = binding.applicationContext
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "start" -> handleStart(call, result)
            "log" -> handleLog(call, result)
            "logScreenView" -> handleLogScreenView(call, result)
            "getSessionId" -> result.success(Logger.sessionId)
            "getSessionUrl" -> result.success(Logger.sessionUrl)
            "getDeviceId" -> result.success(Logger.deviceId)
            "createTemporaryDeviceCode" -> handleCreateTemporaryDeviceCode(result)
            "startNewSession" -> {
                Logger.startNewSession()
                result.success(null)
            }
            "getSdkStatus" -> {
                val status = Logger.getSdkStatus()
                result.success(mapOf(
                    "initializationState" to status.initializationState.name,
                    "lastHandshakeTimeMs" to status.lastHandshakeTimeMs,
                    "lastConfigDeliveryTimeMs" to status.lastConfigDeliveryTimeMs,
                ))
            }
            "addField" -> {
                val key = call.argument<String>("key")!!
                val value = call.argument<String>("value")!!
                Logger.addField(key, value)
                result.success(null)
            }
            "removeField" -> {
                val key = call.argument<String>("key")!!
                Logger.removeField(key)
                result.success(null)
            }
            "setEntityId" -> {
                val entityId = call.argument<String>("entityId")!!
                Logger.setEntityId(entityId)
                result.success(null)
            }
            "clearEntityId" -> {
                Logger.clearEntityId()
                result.success(null)
            }
            "logNetworkRequest" -> handleLogNetworkRequest(call, result)
            "logNetworkResponse" -> handleLogNetworkResponse(call, result)
            "startSpan" -> handleStartSpan(call, result)
            "endSpan" -> handleEndSpan(call, result)
            "logReplayScreen" -> handleLogReplayScreen(call, result)
            else -> result.notImplemented()
        }
    }

    private fun handleStart(call: MethodCall, result: MethodChannel.Result) {
        val apiKey = call.argument<String>("apiKey")!!
        val apiUrl = call.argument<String>("apiUrl") ?: "https://api.bitdrift.io"
        val strategyName = call.argument<String>("sessionStrategy") ?: "fixed"
        val sessionStrategy = when (strategyName) {
            "activityBased" -> SessionStrategy.ActivityBased()
            else -> SessionStrategy.Fixed()
        }
        try {
            Logger.start(
                apiKey = apiKey,
                sessionStrategy = sessionStrategy,
                configuration = Configuration(
                    // Flutter provides its own wireframe data via logSessionReplayScreen.
                    sessionReplayConfiguration = null,
                ),
                apiUrl = apiUrl.toHttpUrl(),
                context = context,
            )
            result.success(true)
        } catch (e: Exception) {
            result.success(false)
        }
    }

    private fun handleLog(call: MethodCall, result: MethodChannel.Result) {
        val level = when (call.argument<String>("level")) {
            "trace" -> LogLevel.TRACE
            "debug" -> LogLevel.DEBUG
            "info" -> LogLevel.INFO
            "warning" -> LogLevel.WARNING
            "error" -> LogLevel.ERROR
            else -> LogLevel.INFO
        }
        val message = call.argument<String>("message")!!
        val fields = call.argument<Map<String, String>>("fields") ?: emptyMap()
        Logger.log(level, fields) { message }
        result.success(null)
    }

    private fun handleCreateTemporaryDeviceCode(result: MethodChannel.Result) {
        Logger.createTemporaryDeviceCode { captureResult ->
            when (captureResult) {
                is CaptureResult.Success -> result.success(captureResult.value)
                is CaptureResult.Failure -> result.error(
                    "DEVICE_CODE_ERROR",
                    captureResult.error.message,
                    null,
                )
            }
        }
    }

    private fun handleLogScreenView(call: MethodCall, result: MethodChannel.Result) {
        val screenName = call.argument<String>("screenName")!!
        Logger.logScreenView(screenName)
        result.success(null)
    }

    @Suppress("UNCHECKED_CAST")
    private fun parseHttpUrlPath(map: Map<*, *>?): HttpUrlPath? {
        val value = map?.get("value") as? String ?: return null
        return HttpUrlPath(value, map["template"] as? String)
    }

    @Suppress("UNCHECKED_CAST")
    private fun parseHttpRequestInfo(map: Map<*, *>): HttpRequestInfo =
        HttpRequestInfo(
            method = map["method"] as String,
            host = map["host"] as? String,
            path = parseHttpUrlPath(map["path"] as? Map<*, *>),
            query = map["query"] as? String,
            headers = map["headers"] as? Map<String, String>,
            bytesExpectedToSendCount = (map["bytesExpectedToSendCount"] as? Number)?.toLong(),
            spanId = UUID.fromString(map["spanId"] as String),
            extraFields = (map["extraFields"] as? Map<String, String>) ?: emptyMap(),
        )

    private fun handleLogNetworkRequest(call: MethodCall, result: MethodChannel.Result) {
        val args = call.arguments as Map<*, *>
        Logger.log(parseHttpRequestInfo(args))
        result.success(null)
    }

    @Suppress("UNCHECKED_CAST")
    private fun handleLogNetworkResponse(call: MethodCall, result: MethodChannel.Result) {
        val args = call.arguments as Map<*, *>
        val request = parseHttpRequestInfo(args["request"] as Map<*, *>)

        val responseMap = args["response"] as Map<*, *>
        val httpResult = when (responseMap["result"] as? String) {
            "success" -> HttpResponse.HttpResult.SUCCESS
            "canceled" -> HttpResponse.HttpResult.CANCELED
            else -> HttpResponse.HttpResult.FAILURE
        }
        val response = HttpResponse(
            result = httpResult,
            host = responseMap["host"] as? String,
            path = parseHttpUrlPath(responseMap["path"] as? Map<*, *>),
            query = responseMap["query"] as? String,
            headers = responseMap["headers"] as? Map<String, String>,
            statusCode = (responseMap["statusCode"] as? Number)?.toInt(),
            error = (responseMap["error"] as? String)?.let { Exception(it) },
        )

        val metricsMap = args["metrics"] as? Map<*, *>
        val metrics = metricsMap?.let {
            HttpRequestMetrics(
                requestBodyBytesSentCount = (it["requestBodyBytesSentCount"] as? Number)?.toLong(),
                responseBodyBytesReceivedCount = (it["responseBodyBytesReceivedCount"] as? Number)?.toLong(),
                requestHeadersBytesCount = (it["requestHeadersBytesCount"] as? Number)?.toLong(),
                responseHeadersBytesCount = (it["responseHeadersBytesCount"] as? Number)?.toLong(),
                dnsResolutionDurationMs = (it["dnsResolutionDurationMs"] as? Number)?.toLong(),
                tlsDurationMs = (it["tlsDurationMs"] as? Number)?.toLong(),
                tcpDurationMs = (it["tcpDurationMs"] as? Number)?.toLong(),
                fetchInitializationMs = (it["fetchInitializationMs"] as? Number)?.toLong(),
                responseLatencyMs = (it["responseLatencyMs"] as? Number)?.toLong(),
                protocolName = it["protocolName"] as? String,
            )
        }

        Logger.log(
            HttpResponseInfo(
                request = request,
                response = response,
                durationMs = (args["durationMs"] as Number).toLong(),
                metrics = metrics,
                extraFields = (args["extraFields"] as? Map<String, String>) ?: emptyMap(),
            ),
        )
        result.success(null)
    }

    private fun handleStartSpan(call: MethodCall, result: MethodChannel.Result) {
        val name = call.argument<String>("name")!!
        val level = when (call.argument<String>("level")) {
            "trace" -> LogLevel.TRACE
            "debug" -> LogLevel.DEBUG
            "warning" -> LogLevel.WARNING
            "error" -> LogLevel.ERROR
            else -> LogLevel.INFO
        }
        val fields = call.argument<Map<String, String>>("fields") ?: emptyMap()
        val span = Logger.startSpan(name, level, fields)
        if (span != null) {
            val spanId = span.hashCode().toString()
            activeSpans[spanId] = span
            result.success(spanId)
        } else {
            result.success(null)
        }
    }

    private fun handleEndSpan(call: MethodCall, result: MethodChannel.Result) {
        val spanId = call.argument<String>("spanId")!!
        val success = call.argument<Boolean>("success") ?: true
        val span = activeSpans.remove(spanId)
        if (span != null) {
            if (success) span.end(SpanResult.SUCCESS) else span.end(SpanResult.FAILURE)
        }
        result.success(null)
    }

    private fun handleLogReplayScreen(call: MethodCall, result: MethodChannel.Result) {
        val screenBytes = call.argument<ByteArray>("screen")!!
        val duration = call.argument<Double>("duration") ?: 0.0
        // logSessionReplayScreen is on the internal IInternalLogger interface.
        // No public API exists for this yet — requires @file:Suppress to access.
        // Track: make logSessionReplayScreen public on ILogger or Capture.Logger.
        try {
            val logger = Capture.logger() as? IInternalLogger
            if (logger != null) {
                val fields = arrayOf(
                    Field("screen", screenBytes.toFieldValue()),
                )
                logger.logSessionReplayScreen(fields, Duration.ZERO)
            }
            result.success(null)
        } catch (e: Exception) {
            result.error("REPLAY_ERROR", e.message, null)
        }
    }

    companion object {
        private val activeSpans = mutableMapOf<String, Span>()
    }
}
